p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
+// an inbound connection while the router is also dialing an outbound
+// connection, which will cause the dialer to eventually error (when attempting
+// to mark the peer as connected). This also avoids race conditions where
+// multiple goroutines may end up dialing a peer if an incoming connection was
+// briefly accepted and disconnected while we were also dialing.
+type peerManager struct {
+	mtx           sync.Mutex
+	store         *peerStore
+	dialing       map[NodeID]bool
+	connected     map[NodeID]bool
+	subscriptions map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
+}
+
+// newPeerManager creates a new peer manager.
+func newPeerManager(store *peerStore) *peerManager {
+	return &peerManager{
+		store:         store,
+		dialing:       map[NodeID]bool{},
+		connected:     map[NodeID]bool{},
+		subscriptions: map[*PeerUpdatesCh]*PeerUpdatesCh{},
+	}
+}
+
+// Add adds a peer to the manager, given as an address. If the peer already
+// exists, the address is added to it.
+func (m *peerManager) Add(address PeerAddress) error {
+	if err := address.Validate(); err != nil {
+		return err
+	}
+	peerID := address.NodeID()
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	peer, err := m.store.Get(peerID)
+	if err != nil {
+		return err
+	}
+	if peer == nil {
+		peer = newPeerInfo(peerID)
+	}
+	if peer.AddAddress(address) {
+		return m.store.Set(peer)
+	}
+	return nil
 }
 
-// newPeerStore creates a new peer store.
-func newPeerStore() *peerStore {
-	return &peerStore{
-		peers:   map[NodeID]*peerInfo{},
-		claimed: map[NodeID]bool{},
+// Subscribe subscribes to peer updates. The caller must consume the peer
+// updates in a timely fashion and close the subscription when done, since
+// delivery is guaranteed and will block peer connection/disconnection
+// otherwise.
+func (m *peerManager) Subscribe() *PeerUpdatesCh {
+	// FIXME: We may want to use a size 1 buffer here. When the router
+	// broadcasts a peer update it has to loop over all of the
+	// subscriptions, and we want to avoid blocking and waiting for a
+	// context switch before continuing to the next subscription. This also
+	// prevents tail latencies from compounding across updates. We also want
+	// to make sure the subscribers are reasonably in sync, so it should be
+	// kept at 1. However, this should be benchmarked first.
+	peerUpdates := NewPeerUpdates(make(chan PeerUpdate))
+	m.mtx.Lock()
+	m.subscriptions[peerUpdates] = peerUpdates
+	m.mtx.Unlock()
+
+	go func() {
+		<-peerUpdates.Done()
+		m.mtx.Lock()
+		delete(m.subscriptions, peerUpdates)
+		m.mtx.Unlock()
+	}()
+	return peerUpdates
+}
+
+// broadcast broadcasts a peer update to all subscriptions. The caller must
+// already hold the mutex lock. This means the mutex is held for the duration
+// of the broadcast, which we want to make sure all subscriptions receive all
+// updates in the same order.
+//
+// FIXME: Consider using more fine-grained mutexes here, and/or a channel to
```

Comment:
you mean some subscriptions are more important than others?

Reply:
I think broadcasting order can be guaranteed but not receiving order assuming we use buffered channels. If the channels are not buffered, then order should be guaranteed.

Reply:
Not really. Ordering within each subscription will be guaranteed regardless of whether they're buffered or not, and I don't see a point in ensuring scheduling ordering of individual subscribers since reactors should not have implicit dependencies on each other.

However, this comment is really more about mutex contention. We call `broadcast()` while holding the global `peerManager.mtx` mutex, which also blocks e.g. peers being added/removed, but `broadcast()` doesn't care what happens to peers, only what happens to `peerManager.subscriptions`. So we could have e.g. a separate mutex for subscriptions, and allow peer changes to take place concurrently with broadcasting a single update.

Reply:
It seems advantageous to have a separate mutex for peer management and subscription management as you've pointed out 👍